core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -833,8 +833,10 @@ private[spark] class BlockManager(
     logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
 
     // Either we're storing bytes and we asynchronously started replication, or we're storing
-    // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
+    // values and need to serialize and replicate them now.
+    // Should not replicate the block if its StorageLevel is StorageLevel.NONE or
+    // putting it to local is failed.
+    if (!putBlockInfo.isFailed && putLevel.replication > 1) {
```

Comment:
Also as I see the code path of doing rdd.Cache , the CacheManager first check if block can be unrolled safely , and then only it tries to call blockManager.putArray(..) which returns the updatedBlocks object. If block is not unrolled safely , CacheManager wont cache the block. So I do not think PR will de-optimize existing code path , as in RDD partitions case, once block is unrolled safely , CacheManager store the block and get the updatedBlocks back, and in this case my fix wont impact anything and replication will happen if block is unrolled safely.  @squito do let me know if this is correct.